File: src/main/php/Easy/PlusOne.php
<?php

declare(strict_types=1);

namespace Src\Easy;

class PlusOne
{
    public function transformPlusOne(array $digits): array
    {
        $result = $digits;
        $remainder = 1;
        for ($index = count($digits) - 1; $index >= 0; $index--) {
            $num = $result[$index] + $remainder;
            if ($num <= 9) {
                $result[$index] = $num;

                return $result;
            }
            $result[$index] = 0;
            $remainder = 1;
        }
        if ($remainder > 0) {
            array_unshift($result, 1);
        }

        return $result;
    }

    public function solve(array $nums): array
    {
        return $this->transformPlusOne($nums);
    }
}

File: lib/easy/plus_one.dart
List<int> plusOne(List<int> digits) {
  List<int> result = digits;
  int index = digits.length - 1;
  int remainder = 1;
  while (index >= 0) {
    int sum = result[index] + remainder;
    if (sum <= 9) {
      result[index] = sum;
      return result;
    }
    result[index] = 0;
    remainder = 1;
    index -= 1;
  }
  if (remainder > 0) {
    result.insert(0, 1);
  }
  return result;
}
